File: 18_4sum/18_4sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        length = len(nums)
        if length<4:
            return []
        nums.sort()
        rtn = []
        for i in range(0, length-3):
            if i>0 and nums[i]==nums[i-1]:
                continue
            if nums[i]*4>target:
                break
            for j in range(i+1, length-2):
                if j>i+1 and nums[j]==nums[j-1]:
                    continue
                if nums[i]+nums[j]+nums[j+1]+nums[j+2]>target:
                    break
                l = j+1
                r = length-1
                if nums[i]+nums[j]+nums[l]+nums[l+1]>target:
                    break
                while l<r:
                    val = nums[i]+nums[j]+nums[l]+nums[r]
                    if val==target:
                        rtn.append([nums[i], nums[j], nums[l], nums[r]])
                        l = l+1
                        r = r-1
                        while l>j+1 and l<length-2 and nums[l]==nums[l-1]:
                            l = l+1
                        while r<length-1 and r>=j+i+2 and nums[r]==nums[r+1]:
                            r = r-1
                    elif val<target:
                        l = l+1
                    else:
                        r = r-1
        return rtn

File: 18_4sum/18_4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        
        int len = nums.size();
        int l,r,sum;
        vector<vector<int>> res;
        
        if (len < 4) return res;
        
        sort(nums.begin(), nums.end());
        
        for (int i = 0; i < len-3; i++)
        {
            if (i>0 && nums[i]==nums[i-1]) continue;
            if (nums[i]+nums[i+1]+nums[i+2]+nums[i+3]>target) break;
            for (int j = i+1; j < len-2; j++)
            {
                if (j>i+1 && nums[j]==nums[j-1]) continue;
                l = j + 1;
                r = len - 1;
                if (nums[i]+nums[j]+nums[l]+nums[l+1]>target) break;
                while (l<r)
                {
                    sum = nums[i] + nums[j] + nums[l] + nums[r];
                    if (sum == target)
                    {
                        vector<int> tmp = {nums[i], nums[j], nums[l], nums[r]};
                        res.push_back(tmp);
                        l++;
                        r--;
                    }
                    else if (sum < target) l++;
                    else r--;
                    while (l>j+1 && l<len-2 && nums[l]==nums[l-1]) l++;
                    while (r<len-1 && r>=j+2 && nums[r]==nums[r+1]) r--;
                }
            }
        }
        
        return res;
        
    }
};